### PR TITLE
Add `s` (CPU) & `m+t4` (GPU) generic machine types

### LINKED ISSUE #328: Consider adding small/micro machine sizes
Currently the machine sizes start at (and default to) `m`. This is fine, and a sane default for the intended use case, but not everyone needs (to pay for) 8 cores and 16GB of RAM to do something basic (like prototyping DVC/SSH features).

Right now the user can just manually specify something like `aws + t2.micro` ourselves but it defeats the purpose of having a generalized solution if the user has to look up the cross-cloud equivalent sizes in the event they want/need to test something against a different cloud provider.

### PR DESCRIPTION
**Pull request by 0x2b3bfa0**
Closes #328. Not tested nor willing to, given the priority of this enhancement; @casperdcl surely will. 😏

### CODE REVIEW

File: docs/resources/task.md [hunk 1]
```diff
@@ -122,9 +122,11 @@ The Iterative Provider offers some common machine types (medium, large, and extr
 
 | Type      | Minimum CPU cores | Minimum RAM | GPU                 |
 | :-------- | ----------------: | ----------: | :------------------ |
+| `s`       |                 1 |        1 GB | -                   |
 | `m`       |                 8 |       16 GB | -                   |
 | `l`       |                32 |       64 GB | -                   |
 | `xl`      |                64 |      128 GB | -                   |
+| `s+t4`    |                 1 |        1 GB | -                   |
```
Comment (casperdcl) [inline docs/resources/task.md:129]:
```suggestion
| `s+t4`    |                 1 |        1 GB | 1 NVIDIA Tesla T4   |
```

Comment (0x2b3bfa0) [reply docs/resources/task.md]: Whoops! Thanks! 🙏

**Review by casperdcl — COMMENTED** [review]
I don't grok the typespec, so actually may as well discuss before adding new weird things.

- Currently: CPU_num(`s/m/l/xl`)+GPU_type(`<series>`)
- Alternative (1): don't need to fill in all combos for all providers
  * e.g. AWS & Azure don't really have `s+t4` so replace the label with `m+t4` just for them
- Alternative (2): CPU_num(`s/m/l/xl`)+GPU_mem(`s/m/l/xl`)
  * because that's what most people care about?
  * `s+s` can thus have the cheapest GPU combo on all providers (rather than restrict to T4 GPUs)
- Alternative (3): CPU_num(int)+GPU_mem_GB(int)
  * explicit minimums!

P.S. cheapest GPU is really:

- AWS: T4 (`g4dn.xlarge`), would call this `m+t4`
- Azure: T4 (`Standard_NC4as_T4_v3`), would call this `m+t4`, but will be A10 (`Standard_NV6ads_A10_v5`) aka `s+a10` soon ([currently in preview](https://docs.microsoft.com/en-us/azure/virtual-machines/nva10v5-series))
- GCP: K80 (`n1-custom-2-12288+nvidia-tesla-k80*1`), would call this `s+k80`

P.P.S. isn't https://azure.microsoft.com/en-us/pricing/calculator better than https://azure.microsoft.com/en-us/pricing/vm-selector to compare pricing info?

P.P.P.S isn't https://cloud.google.com/products/calculator better than https://cloud.google.com/compute/docs/machine-types to compare pricing info?

**Comment by 0x2b3bfa0** [comment]
> _I don't grok the typespec, so actually may as well discuss before adding new weird things._

* https://github.com/iterative/terraform-provider-iterative/issues/158

> `(?:P\.){2,3}S\.`

I find https://azure.microsoft.com/en-us/pricing/vm-selector and https://cloud.google.com/compute/vm-instance-pricing reasonably useful.

File: docs/guides/generic-machine-types.md [hunk 2]
```diff
@@ -7,17 +7,19 @@ subcategory: Development
...
+| Type      | [`aws`]       | [`az`]                 | [`gcp`]                                         | [`k8s`]                                              |
+| :-------- | :------------ | :--------------------- | :---------------------------------------------- | :--------------------------------------------------- |
+| `s`       | `t2.micro`    | `Standard_B1s`         | `g1-small`                                      | `cpu: 1`<br>`memory: 1G`                             |
+| `m`       | `m5.2xlarge`  | `Standard_F8s_v2`      | `e2-custom-8-32768`                             | `cpu: 8`<br>`memory: 32G`                            |
+| `l`       | `m5.8xlarge`  | `Standard_F32s_v2`     | `e2-custom-32-131072`                           | `cpu: 32`<br>`memory: 128G`                          |
+| `xl`      | `m5.16xlarge` | `Standard_F64s_v2`     | `n2-custom-64-262144`                           | `cpu: 64`<br>`memory: 256G`                          |
+| `s+t4`    | `g4dn.xlarge` | `Standard_NC4as_T4_v3` | `custom-4-16384`<br>1 `nvidia-tesla-t4`         | `cpu: 1`<br>`memory: 1G`<br>1 `nvidia-tesla-t4`      |
```
Comment (casperdcl) [inline docs/guides/generic-machine-types.md:16]:
```suggestion
| `m+t4`    | `g4dn.xlarge` | `Standard_NC4as_T4_v3` | `n1-standard-4`<br>1 `nvidia-tesla-t4`                 | `cpu: 4`<br>`memory: 16G`<br>1 `nvidia-tesla-t4`      |
```

or even

```suggestion
| `s+k80`   |             - |                      - | `n1-highmem-2`<br>1 `nvidia-tesla-k80`                | `cpu: 2`<br>`memory: 12G`<br>1 `nvidia-tesla-k80`     |
| `m+t4`    | `g4dn.xlarge` | `Standard_NC4as_T4_v3` | `n1-standard-4`<br>1 `nvidia-tesla-t4`                 | `cpu: 4`<br>`memory: 16G`<br>1 `nvidia-tesla-t4`      |
```

File: docs/resources/task.md [hunk 3]
```diff
@@ -122,9 +122,11 @@ The Iterative Provider offers some common machine types (medium, large, and extr
 
 | Type      | Minimum CPU cores | Minimum RAM | GPU                 |
 | :-------- | ----------------: | ----------: | :------------------ |
+| `s`       |                 1 |        1 GB | -                   |
 | `m`       |                 8 |       16 GB | -                   |
 | `l`       |                32 |       64 GB | -                   |
 | `xl`      |                64 |      128 GB | -                   |
+| `s+t4`    |                 1 |        1 GB | 1 NVIDIA Tesla T4   |
```
Comment (casperdcl) [inline docs/resources/task.md:129]:
```suggestion
| `m+t4`    |                 4 |       16 GB | 1 NVIDIA Tesla T4   |
```

or even

```suggestion
| `s+k80`   |                 2 |       12 GB | 1 NVIDIA Tesla K80  |
| `m+t4`    |                 4 |       16 GB | 1 NVIDIA Tesla T4   |
```

File: task/aws/resources/resource_launch_template.go [hunk 4]
```diff
@@ -55,9 +55,11 @@ func (l *LaunchTemplate) Create(ctx context.Context) error {
 
 	size := l.Attributes.Size.Machine
 	sizes := map[string]string{
+		"s":       "t2.micro",
 		"m":       "m5.2xlarge",
 		"l":       "m5.8xlarge",
 		"xl":      "m5.16xlarge",
+		"s+t4":    "g4dn.xlarge",
```
Comment (casperdcl) [inline task/aws/resources/resource_launch_template.go:62]:
```suggestion
		"m+t4":    "g4dn.xlarge",
```

File: task/az/resources/resource_virtual_machine_scale_set.go [hunk 5]
```diff
@@ -105,9 +105,11 @@ func (v *VirtualMachineScaleSet) Create(ctx context.Context) error {
 
 	size := v.Attributes.Size.Machine
 	sizes := map[string]string{
+		"s":       "Standard_B1s",
 		"m":       "Standard_F8s_v2",
 		"l":       "Standard_F32s_v2",
 		"xl":      "Standard_F64s_v2",
+		"s+t4":    "Standard_NC4as_T4_v3",
```
Comment (casperdcl) [inline task/az/resources/resource_virtual_machine_scale_set.go:112]:
```suggestion
		"m+t4":    "Standard_NC4as_T4_v3",
```

File: task/gcp/resources/resource_instance_template.go [hunk 6]
```diff
@@ -67,9 +67,11 @@ func (i *InstanceTemplate) Create(ctx context.Context) error {
 
 	size := i.Attributes.Size.Machine
 	sizes := map[string]string{
+		"s":       "g1-small",
 		"m":       "e2-custom-8-32768",
 		"l":       "e2-custom-32-131072",
 		"xl":      "n2-custom-64-262144",
+		"s+t4":    "g1-small+nvidia-tesla-t4*1",
```
Comment (casperdcl) [inline task/gcp/resources/resource_instance_template.go:74]:
```suggestion
		"s+k80":   "n1-highmem-2+nvidia-tesla-k80*1",
		"m+t4":    "n1-standard-4+nvidia-tesla-t4*1",
```

File: task/k8s/resources/resource_job.go [hunk 7]
```diff
@@ -58,9 +58,11 @@ type Job struct {
 func (j *Job) Create(ctx context.Context) error {
 	size := j.Attributes.Task.Size.Machine
 	sizes := map[string]string{
+		"s":       "1-1000",
 		"m":       "8-32000",
 		"l":       "32-128000",
 		"xl":      "64-256000",
+		"s+t4":    "1-1000+nvidia-tesla-t4*1",
```
Comment (casperdcl) [inline task/k8s/resources/resource_job.go:65]:
```suggestion
		"s+k80":   "2-12000+nvidia-tesla-k80*1",
		"m+t4":    "4-16000+nvidia-tesla-t4*1",
```

**Comment by 0x2b3bfa0** [comment]
@casperdcl, from what I inderstand from your last suggestions, they would be introducing uneven machine types across providers? 🤔

**Comment by casperdcl** [comment]
[this change summarises this iteration](https://github.com/iterative/terraform-provider-iterative/pull/525#discussion_r855962209)

File: docs/resources/task.md (hunk 3 above)
Comment (0x2b3bfa0) [reply docs/resources/task.md:129]: > _On some providers_

Here is where I stop reading. 🐉

Users that are dumb enough not to check their provider's documentation should be able to use our small, curated set of defaults on every provider.

If they have to figure out why some allegedly generic machine type doesn't work in their cloud provider, something is wrong with us.

File: task/aws/resources/resource_launch_template.go [hunk 8]
```diff
@@ -55,9 +55,11 @@ func (l *LaunchTemplate) Create(ctx context.Context) error {
...
 		"m+k80":   "p2.xlarge",
 		"l+k80":   "p2.8xlarge",
 		"xl+k80":  "p2.16xlarge",
```
Comment (casperdcl) [inline task/aws/resources/resource_launch_template.go:65]: also, if none of these `short` -> `long` mapping succeed && the result still looks like a short string, we should log a warning mentioning https://registry.terraform.io/providers/iterative/iterative/latest/docs/guides/generic-machine-types

File: docs/resources/task.md (hunk 3 above)
Comment (0x2b3bfa0) [reply docs/resources/task.md:129]: I would provide only a small set of “common” machine types for demonstration purposes; it's “fine” 🔥🐶 to add `m+t4` for demonstration & testing purposes. 

For now at least, users should be responsible for navigating their cloud provider's documentation if they want other instance types.

File: task/aws/resources/resource_launch_template.go (hunk 8 above)
Comment (0x2b3bfa0) [reply task/aws/resources/resource_launch_template.go:65]: I still believe that this kind of behavior, while convenient, is [not a priority](https://github.com/iterative/terraform-provider-iterative/pull/525#discussion_r856149156) for now. 🤔

File: task/gcp/resources/resource_instance_template.go [hunk 9]
```diff
@@ -67,9 +67,12 @@ func (i *InstanceTemplate) Create(ctx context.Context) error {
 
 	size := i.Attributes.Size.Machine
 	sizes := map[string]string{
+		"s":       "g1-small",
 		"m":       "e2-custom-8-32768",
 		"l":       "e2-custom-32-131072",
 		"xl":      "n2-custom-64-262144",
+		"s+k80":   "n1-highmem-2+nvidia-tesla-k80*1",
```
Comment (0x2b3bfa0) [inline task/gcp/resources/resource_instance_template.go:74]: Argh!

File: task/k8s/resources/resource_job.go [hunk 10]
```diff
@@ -58,9 +58,12 @@ type Job struct {
 func (j *Job) Create(ctx context.Context) error {
 	size := j.Attributes.Task.Size.Machine
 	sizes := map[string]string{
+		"s":       "1-1000",
 		"m":       "8-32000",
 		"l":       "32-128000",
 		"xl":      "64-256000",
+		"s+k80":   "2-12000+nvidia-tesla-k80*1",
```
Comment (0x2b3bfa0) [inline task/k8s/resources/resource_job.go:65]: Argh!

File: docs/guides/generic-machine-types.md [hunk 11]
```diff
@@ -7,17 +7,20 @@ subcategory: Development
...
+| Type      | [`aws`]       | [`az`]                 | [`gcp`]                                         | [`k8s`]                                              |
+| :-------- | :------------ | :--------------------- | :---------------------------------------------- | :--------------------------------------------------- |
+| `s`       | `t2.micro`    | `Standard_B1s`         | `g1-small`                                      | `cpu: 1`<br>`memory: 1G`                             |
+| `m`       | `m5.2xlarge`  | `Standard_F8s_v2`      | `e2-custom-8-32768`                             | `cpu: 8`<br>`memory: 32G`                            |
+| `l`       | `m5.8xlarge`  | `Standard_F32s_v2`     | `e2-custom-32-131072`                           | `cpu: 32`<br>`memory: 128G`                          |
+| `xl`      | `m5.16xlarge` | `Standard_F64s_v2`     | `n2-custom-64-262144`                           | `cpu: 64`<br>`memory: 256G`                          |
+| `s+k80`   | -             | -                      | `n1-highmem-2`<br>1 `nvidia-tesla-k80`          | `cpu: 2`<br>`memory: 12G`<br>1 `nvidia-tesla-k80`    |
```
Comment (0x2b3bfa0) [inline docs/guides/generic-machine-types.md:16]: Argh!

**Comment by 0x2b3bfa0** [comment]
@casperdcl, it looks like you forgot some `s+k80` there?

**Comment by 0x2b3bfa0** [comment]
> _Alternative (3): CPU_num(int)+GPU_mem_GB(int)_
> _explicit minimums!_

See also https://github.com/iterative/terraform-provider-iterative/issues/158#issuecomment-934826522 → Turning it into a code problem

**Comment by 0x2b3bfa0** [comment]
The original XPD alpha (what became `task` in a few weeks) had this implemented.

**Pull request merged**